consensus/src/merkle.rs

```diff
@@ -0,0 +1,223 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+//
+// Copyright (c) DUSK NETWORK. All rights reserved.
+
+use dusk_merkle::{Aggregate, Tree};
+use sha3::{Digest, Sha3_256};
+
+pub const ARITY: usize = 2;
+
+#[derive(Clone, Copy)]
+pub struct Hash([u8; 32]);
+
+impl<I: Into<[u8; 32]>> From<I> for Hash {
+    fn from(bytes: I) -> Self {
+        Self(bytes.into())
+    }
+}
+
+pub const EMPTY_NODE: Hash = Hash([0; 32]);
+
+impl<const H: usize> Aggregate<H, ARITY> for Hash {
+    const EMPTY_SUBTREES: [Self; H] = [EMPTY_NODE; H];
+
+    /// old golang implementation duplicates the missing leaf
+    ///
+    /// E.g: for (H=3, A=2)
+    ///
+    /// if you have
+    /// 1 2 3 4 5
+    /// it will create duplicates
+    /// 1 2 3 4 5 5 5 5
+    ///
+    /// if you have
+    /// 1 2 3 4 5 6
+    /// it creates
+    /// 1 2 3 4 5 6 5 6
+    fn aggregate<'a, I>(items: I) -> Self
+    where
+        Self: 'a,
+        I: Iterator<Item = &'a Self>,
+    {
+        let mut hasher = Sha3_256::new();
+        let mut prev = &[0u8; 32];
+        for item in items {
+            if item.0 == EMPTY_NODE.0 {
+                hasher.update(prev);
+            } else {
+                hasher.update(item.0);
+                prev = &item.0;
+            };
+        }
+        Self::from(hasher.finalize())
+    }
+}
+
+struct BinaryMerkle<const H: usize> {
+    tree: Tree<Hash, H, ARITY>,
+}
+
+impl<const H: usize> BinaryMerkle<H> {
+    const fn new() -> Self {
+        Self { tree: Tree::new() }
+    }
+
+    fn insert<N: Into<Hash>>(&mut self, val: N) {
+        let val: [u8; 32] = Into::<Hash>::into(val).0;
+        self.tree.insert(self.tree.len(), val);
+    }
+
+    /// Returns the root of the smallest inner tree
+    fn shrunken_root(&self) -> [u8; 32] {
+        if self.tree.is_empty() {
+            return EMPTY_NODE.0;
+        }
+
+        let inner_tree_height =
+            (self.tree.len() as f64 - 1f64).sqrt().floor() as usize + 1;
+
+        if inner_tree_height == H {
+            return self.tree.root().0;
+        }
+
+        let opening = self
+            .tree
+            .opening(0)
```

Comment:
We insert leaves at a specific position so there could be a case where there is no leaf at pos=0 but the tree is still not empty. Use the walker to retrieve any leaf instead